zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java

```diff
@@ -1711,6 +1724,50 @@ private void getInterpreterSettings(NotebookSocket conn, AuthenticationInfo subj
     conn.send(serializeMessage(new Message(OP.INTERPRETER_SETTINGS)
             .put("interpreterSettings", availableSettings)));
   }
-
+  
+  private void switchConnectionToWatcher(NotebookSocket conn, Message messagereceived)
+      throws IOException {
+    if (!isSessionAllowedToSwitchToWatcher(conn)) {
+      LOG.error("Cannot switch this client to watcher, invalid security key");
+      return;
+    }
+    LOG.info("Going to add {} to watcher socket", conn);
+    // add the connection to the watcher.
+    if (watcherSockets.contains(conn)) {
+      LOG.info("connection alrerady present in the watcher");
+      return;
+    }
+    watcherSockets.add(conn);
+    
+    // remove this connection from regular zeppelin ws usage.
+    removeConnectionFromAllNote(conn);
+    connectedSockets.remove(conn);
+    removeUserConnection(conn.getUser(), conn);
+  }
+  
+  private boolean isSessionAllowedToSwitchToWatcher(NotebookSocket session) {
+    String watcherSecurityKey = session.getRequest().getHeader(WatcherSecurityKey.HTTP_HEADER);
+    return !(StringUtils.isBlank(watcherSecurityKey)
+        || !watcherSecurityKey.equals(WatcherSecurityKey.getKey()));
```

Comment:
yeah i see, makes sense. generally to make it more readable maybe we could do something like:
```
String a = "a";
String b = "b";

(!StringUtils.isBlank(a) && a.equals(b))) // -> false

(!StringUtils.isBlank(a) && a.equals(a))) // -> true
```
which is quite straightforward to understand from first read